File: typescript/tests/pii.test.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import { afterEach, describe, expect, it } from 'vitest';
import {
  DEFAULT_SANITIZE_FIELDS,
  _setHashFnForTest,
  getPiiRules,
  registerPiiRule,
  replacePiiRules,
  resetPiiRulesForTests,
  sanitize,
  sanitizePayload,
} from '../src/pii';

afterEach(() => resetPiiRulesForTests());

describe('sanitize (backwards-compat)', () => {
  it('redacts default PII fields', () => {
    const obj: Record<string, unknown> = { password: 'secret', status: 200 };
    sanitize(obj);
    expect(obj['password']).toBe('***');
    expect(obj['status']).toBe(200);
  });

  it('redacts extra fields', () => {
    const obj: Record<string, unknown> = { my_field: 'val', other: 'ok' };
    sanitize(obj, ['my_field']);
    expect(obj['my_field']).toBe('***');
    expect(obj['other']).toBe('ok');
  });

  it('is case-insensitive', () => {
    const obj: Record<string, unknown> = { PASSWORD: 'x', Token: 'y' };
    sanitize(obj);
    expect(obj['PASSWORD']).toBe('***');
    expect(obj['Token']).toBe('***');
  });

  it('DEFAULT_SANITIZE_FIELDS includes standard keys', () => {
    expect(DEFAULT_SANITIZE_FIELDS).toContain('password');
    expect(DEFAULT_SANITIZE_FIELDS).toContain('token');
    expect(DEFAULT_SANITIZE_FIELDS).toContain('api_key');
  });
});

describe('PIIRule registry', () => {
  it('starts empty', () => {
    expect(getPiiRules()).toHaveLength(0);
  });

  it('registerPiiRule appends', () => {
    registerPiiRule({ path: 'user.email', mode: 'redact' });
    expect(getPiiRules()).toHaveLength(1);
    expect(getPiiRules()[0].path).toBe('user.email');
  });

  it('replacePiiRules replaces all', () => {
    registerPiiRule({ path: 'a', mode: 'drop' });
    replacePiiRules([{ path: 'b', mode: 'hash' }]);
    expect(getPiiRules()).toHaveLength(1);
    expect(getPiiRules()[0].path).toBe('b');
  });

  it('getPiiRules returns a copy', () => {
    registerPiiRule({ path: 'x', mode: 'redact' });
    const copy = getPiiRules();
    copy.push({ path: 'y', mode: 'drop' });
    expect(getPiiRules()).toHaveLength(1);
  });
});

describe('sanitizePayload — redact mode', () => {
  it('redacts a matching top-level key', () => {
    registerPiiRule({ path: 'secret', mode: 'redact' });
    const obj: Record<string, unknown> = { secret: 'abc', keep: 'me' };
    sanitizePayload(obj);
    expect(obj['secret']).toBe('***');
    expect(obj['keep']).toBe('me');
  });
});

describe('sanitizePayload — drop mode', () => {
  it('removes the key', () => {
    registerPiiRule({ path: 'token', mode: 'drop' });
    const obj: Record<string, unknown> = { token: 'jwt', other: 1 };
    sanitizePayload(obj);
    expect('token' in obj).toBe(false);
    expect(obj['other']).toBe(1);
  });
});

describe('sanitizePayload — truncate mode', () => {
  it('truncates to truncateTo chars + "..."', () => {
    registerPiiRule({ path: 'note', mode: 'truncate', truncateTo: 5 });
    const obj: Record<string, unknown> = { note: 'hello world' };
    sanitizePayload(obj);
    expect(obj['note']).toBe('hello...');
  });

  it('does not truncate short values', () => {
    registerPiiRule({ path: 'note', mode: 'truncate', truncateTo: 20 });
    const obj: Record<string, unknown> = { note: 'hi' };
    sanitizePayload(obj);
    expect(obj['note']).toBe('hi');
  });

  it('uses default truncateTo=8 when not specified', () => {
    registerPiiRule({ path: 'note', mode: 'truncate' });
    const obj: Record<string, unknown> = { note: 'hello world 123456789' };
    sanitizePayload(obj);
    expect(String(obj['note'])).toMatch(/^.{8}\.\.\.$/);
  });
});

describe('sanitizePayload — hash mode', () => {
  it('hashes the value to a hex string', () => {
    registerPiiRule({ path: 'id', mode: 'hash' });
    const obj: Record<string, unknown> = { id: 'user-123' };
    sanitizePayload(obj);
    expect(typeof obj['id']).toBe('string');
    expect(String(obj['id'])).toMatch(/^[0-9a-f]{12}$/);
  });

  it('is deterministic — same input → same output', () => {
    registerPiiRule({ path: 'id', mode: 'hash' });
    const obj1: Record<string, unknown> = { id: 'user-abc' };
    const obj2: Record<string, unknown> = { id: 'user-abc' };
    sanitizePayload(obj1);
    sanitizePayload(obj2);
    expect(obj1['id']).toBe(obj2['id']);
  });

  it('falls back to [HASHED] when crypto is unavailable', () => {
    _setHashFnForTest(() => {
      throw new Error('no crypto');
    });
    registerPiiRule({ path: 'id', mode: 'hash' });
    const obj: Record<string, unknown> = { id: 'user-123' };
    sanitizePayload(obj);
    expect(obj['id']).toBe('***');
  });
});

describe('sanitizePayload — array values', () => {
  it('applies rules recursively to array items using wildcard path', () => {
    // Rule: 'items.*.id' targets the 'id' field inside each array element of 'items'
    registerPiiRule({ path: 'items.*.id', mode: 'drop' });
    const obj: Record<string, unknown> = {
      items: [
        { id: 'u1', name: 'Alice' },
        { id: 'u2', name: 'Bob' },
      ],
    };
    sanitizePayload(obj);
    const items = obj['items'] as Array<Record<string, unknown>>;
    expect('id' in items[0]).toBe(false);
    expect('id' in items[1]).toBe(false);
    expect(items[0]['name']).toBe('Alice');
    expect(items[1]['name']).toBe('Bob');
  });
});

describe('sanitizePayload — default sensitive keys', () => {
  it('redacts default sensitive keys even without a rule', () => {
    const obj: Record<string, unknown> = { password: 'secret', name: 'Alice' };
    sanitizePayload(obj);
    expect(obj['password']).toBe('***');
    expect(obj['name']).toBe('Alice');
  });

  it('extraFields parameter also gets redacted', () => {
    const obj: Record<string, unknown> = { ssn: '123', custom: 'val' };
    sanitizePayload(obj, ['custom']);
    expect(obj['ssn']).toBe('***');
    expect(obj['custom']).toBe('***');
  });

  it('does not double-redact keys already handled by a rule', () => {
    registerPiiRule({ path: 'token', mode: 'drop' });
    const obj: Record<string, unknown> = { token: 'jwt', other: 1 };
    sanitizePayload(obj);
    expect('token' in obj).toBe(false); // rule says drop
  });

  it('uses exact-match default key detection instead of substring matching', () => {
    const obj: Record<string, unknown> = {
      author_id: 'safe-author',
      spinning_wheel: 'safe-spin',
      glassness: 'safe-word',
    };
    sanitizePayload(obj);
    expect(obj['author_id']).toBe('safe-author');
    expect(obj['spinning_wheel']).toBe('safe-spin');
    expect(obj['glassness']).toBe('safe-word');
  });
});

describe('DEFAULT_SANITIZE_FIELDS — complete coverage', () => {
  it('includes all expected PII field names (canonical 17-key list)', () => {
    const fields = DEFAULT_SANITIZE_FIELDS;
    expect(fields).toContain('passwd');
    expect(fields).toContain('secret');
    expect(fields).toContain('authorization');
    expect(fields).toContain('cookie');
    expect(fields).toContain('credit_card');
    expect(fields).toContain('ssn');
    expect(fields).toContain('private_key');
    // New keys in the canonical list
    expect(fields).toContain('credential');
    expect(fields).toContain('cvv');
    expect(fields).toContain('pin');
    expect(fields).toContain('account_number');
    expect(fields).toContain('apikey');
    expect(fields).toContain('creditcard');
    expect(fields).toContain('auth');
    // email is intentionally NOT in the default list
    expect(fields).not.toContain('email');
  });
});

describe('sanitizePayload — truncate boundary', () => {
  it('does NOT truncate when value length equals truncateTo exactly', () => {
    // text.length > limit (not >=): value of exactly truncateTo chars should be preserved
    registerPiiRule({ path: 'note', mode: 'truncate', truncateTo: 5 });
    const obj: Record<string, unknown> = { note: 'hello' }; // length=5 == truncateTo=5
    sanitizePayload(obj);
    expect(obj['note']).toBe('hello'); // not 'hello...'
  });

  it('truncates when value length is one more than truncateTo', () => {
    registerPiiRule({ path: 'note', mode: 'truncate', truncateTo: 5 });
    const obj: Record<string, unknown> = { note: 'hello!' }; // length=6 > truncateTo=5
    sanitizePayload(obj);
    expect(obj['note']).toBe('hello...');
  });
});

describe('sanitizePayload — wildcard on object keys', () => {
  it('applies rule to object field matched by wildcard segment', () => {
    // Rule 'users.*.email' should match object like { users: { alice: { email: '...' } } }
    registerPiiRule({ path: 'users.*.email', mode: 'redact' });
    const obj: Record<string, unknown> = {
      users: {
        alice: { email: 'alice@example.com', role: 'admin' },
        bob: { email: 'bob@example.com', role: 'user' },
      },
    };
    sanitizePayload(obj);
    const users = obj['users'] as Record<string, Record<string, unknown>>;
    expect(users['alice']['email']).toBe('***');
    expect(users['bob']['email']).toBe('***');
    expect(users['alice']['role']).toBe('admin'); // unaffected
  });
});

describe('pii — null node passthrough (kills ConditionalExpression on node===null)', () => {
  it('sanitizePayload with null nested value does not throw', () => {
    registerPiiRule({ path: 'user.email', mode: 'redact' });
    const obj: Record<string, unknown> = { user: null };
    expect(() => sanitizePayload(obj)).not.toThrow();
    expect(obj['user']).toBeNull();
  });

  it('sanitizePayload with primitive nested value does not throw', () => {
    registerPiiRule({ path: 'meta.token', mode: 'redact' });
    const obj: Record<string, unknown> = { meta: 42 };
    expect(() => sanitizePayload(obj)).not.toThrow();
    expect(obj['meta']).toBe(42);
  });
});

describe('pii — ruleTargets optimization (kills OptionalChaining + MethodExpression)', () => {
  it('does not redact fields not matching any rule leaf', () => {
    resetPiiRulesForTests();
    registerPiiRule({ path: 'user.email', mode: 'redact' });
    const obj: Record<string, unknown> = { unrelated: 'value', other_field: 'data' };
    sanitizePayload(obj);
    expect(obj['unrelated']).toBe('value');
    expect(obj['other_field']).toBe('data');
  });

  it('only applies rule for the matching leaf path segment', () => {
    resetPiiRulesForTests();
    registerPiiRule({ path: 'a.b.secret', mode: 'redact' });
    const obj: Record<string, unknown> = {
      a: { b: { secret: 'hidden', other: 'visible' } },
    };
    sanitizePayload(obj);
    const aObj = obj['a'] as Record<string, unknown>;
    const bObj = aObj['b'] as Record<string, unknown>;
    expect(bObj['secret']).toBe('***');
    expect(bObj['other']).toBe('visible');
  });
});

describe('pii — ruleTargets dedup guard (kills ArrowFunction + MethodExpression at pii.ts:178)', () => {
  it('hash rule on password prevents default redaction from overriding with [REDACTED]', () => {
    // ruleTargets collects last path segments of rules (via .pop()?.toLowerCase()).
    // If ArrowFunction is mutated → () => undefined, ruleTargets = {undefined}, ruleTargets.has('password') = false
    // → default redaction applies to 'password' even though a rule already handled it → value becomes [REDACTED].
    // If MethodExpression mutates .pop() → undefined, same effect.
    // With correct implementation: ruleTargets = {'password'}, ruleTargets.has('password') = true → default redaction skipped.
    resetPiiRulesForTests();
    registerPiiRule({ path: 'password', mode: 'hash' });
    const obj: Record<string, unknown> = { password: 'hunter2', name: 'Alice' }; // pragma: allowlist secret
    sanitizePayload(obj);
    // With correct code, hash rule handled password — result is a hash, NOT [REDACTED]
    expect(typeof obj['password']).toBe('string');
    expect(obj['password']).not.toBe('***');
    expect(String(obj['password'])).toMatch(/^[0-9a-f]{12}$/); // 12-char hex hash
    expect(obj['name']).toBe('Alice'); // unaffected
  });

  it('truncate rule on password preserves truncated value (not overridden by [REDACTED])', () => {
    // password is a DEFAULT_SANITIZE_FIELD. With ArrowFunction/MethodExpression mutation, ruleTargets.has('password') = false
    // → default redaction overrides the truncated value with [REDACTED].
    // With correct code: ruleTargets.has('password') = true → default redaction skipped → truncated value preserved.
    resetPiiRulesForTests();
    registerPiiRule({ path: 'password', mode: 'truncate', truncateTo: 4 });
    const obj: Record<string, unknown> = { password: 'hunter2', name: 'Bob' }; // pragma: allowlist secret
    sanitizePayload(obj);
    // With correct code: truncated value 'hunt...' NOT '***'
    expect(String(obj['password'])).toBe('hunt...');
    expect(obj['name']).toBe('Bob');
  });
});

// Secret detection tests (_detectSecretInValue, _SECRET_PATTERNS, registerSecretPattern) live in pii.secrets.test.ts

describe('sanitizePayload — obj key update from transformed result (kills line 225)', () => {
  it('updates original obj keys from rule-transformed result', () => {
    registerPiiRule({ path: 'data', mode: 'redact' });
    const obj: Record<string, unknown> = { data: 'secret', name: 'Alice' };
    sanitizePayload(obj);
    expect(obj['data']).toBe('***');
    expect(obj['name']).toBe('Alice');
  });

  it('deletes keys from original obj when rule drops them', () => {
    registerPiiRule({ path: 'remove_me', mode: 'drop' });
    const obj: Record<string, unknown> = { remove_me: 'gone', keep: 'here' };
    sanitizePayload(obj);
    expect('remove_me' in obj).toBe(false);
    expect(obj['keep']).toBe('here');
  });
});

describe('_applyRuleFull — depth limit (kills line 173 depth >= maxDepth branch)', () => {
  afterEach(() => resetPiiRulesForTests());

  it('stops recursing at maxDepth and leaves nested value unredacted', () => {
    registerPiiRule({ path: 'a.b.email', mode: 'redact' });
    const obj: Record<string, unknown> = { a: { b: { email: 'alice@example.com' } } };
    // maxDepth=1: _applyRuleFull reaches depth=1 at 'a' dict and returns without traversing further
    sanitizePayload(obj, [], { maxDepth: 1 });
    // The email is NOT redacted because depth limit was hit before the rule path matched
    const a = obj['a'] as Record<string, unknown>;
    const b = a['b'] as Record<string, unknown>;
    expect(b['email']).toBe('alice@example.com');
  });
});

File: go/pii_test.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

package telemetry

import (
	"strings"
	"testing"
)

const (
	_testAlice   = "alice"
	_testMutated = "mutated"
)

// ── helpers ──────────────────────────────────────────────────────────────────

func resetPII(t *testing.T) {
	t.Helper()
	_resetPIIRules()
	_resetSecretPatterns()
	t.Cleanup(_resetPIIRules)
	t.Cleanup(_resetSecretPatterns)
}

// ── Test 1: disabled returns shallow copy unchanged ───────────────────────────

func TestSanitizePayload_Disabled_ReturnsCopyUnchanged(t *testing.T) {
	resetPII(t)
	payload := map[string]any{
		"password": "notsecret", // pragma: allowlist secret
		"user":     _testAlice,
	}
	result := SanitizePayload(payload, false, 32)
	if result["password"] != "notsecret" { // pragma: allowlist secret
		t.Errorf("expected password unchanged, got %v", result["password"])
	}
	if result["user"] != _testAlice {
		t.Errorf("expected user unchanged, got %v", result["user"])
	}
	// Verify it's a copy, not the same map.
	result["user"] = _testMutated
	if payload["user"] != _testAlice {
		t.Error("expected original map to be unmodified")
	}
}

// ── Test 2: default sensitive key detection ───────────────────────────────────

func TestSanitizePayload_DefaultSensitiveKeys_Redacted(t *testing.T) {
	resetPII(t)
	cases := []string{
		"password", "passwd", "secret", "token", "api_key", "apikey",
		"auth", "authorization", "credential", "private_key", "ssn",
		"credit_card", "creditcard", "cvv", "pin", "account_number",
	}
	for _, key := range cases {
		payload := map[string]any{key: "sensitive-value"}
		result := SanitizePayload(payload, true, 32)
		if result[key] != _piiRedacted {
			t.Errorf("key %q: expected %q, got %v", key, _piiRedacted, result[key])
		}
	}
}

func TestSanitizePayload_NonSensitiveKey_NotRedacted(t *testing.T) {
	resetPII(t)
	payload := map[string]any{"username": "alice", "age": 30}
	result := SanitizePayload(payload, true, 32)
	if result["username"] != "alice" {
		t.Errorf("expected username unchanged, got %v", result["username"])
	}
	if result["age"] != 30 {
		t.Errorf("expected age unchanged, got %v", result["age"])
	}
}

// ── Test 3: custom rule — drop mode ──────────────────────────────────────────

func TestSanitizePayload_CustomRule_Drop(t *testing.T) {
	resetPII(t)
	SetPIIRules([]PIIRule{
		{Path: []string{"card_number"}, Mode: PIIModeDrop},
	})
	payload := map[string]any{
		"card_number": "4111-1111-1111-1111",
		"user":        "bob",
	}
	result := SanitizePayload(payload, true, 32)
	if _, ok := result["card_number"]; ok {
		t.Error("expected card_number to be dropped")
	}
	if result["user"] != "bob" {
		t.Errorf("expected user unchanged, got %v", result["user"])
	}
}

// ── Test 4: custom rule — redact mode ────────────────────────────────────────

func TestSanitizePayload_CustomRule_Redact(t *testing.T) {
	resetPII(t)
	SetPIIRules([]PIIRule{
		{Path: []string{"email"}, Mode: PIIModeRedact},
	})
	payload := map[string]any{"email": "alice@example.com"}
	result := SanitizePayload(payload, true, 32)
	if result["email"] != _piiRedacted {
		t.Errorf("expected %q, got %v", _piiRedacted, result["email"])
	}
}

// ── Test 5: custom rule — hash mode ──────────────────────────────────────────

func TestSanitizePayload_CustomRule_Hash(t *testing.T) {
	resetPII(t)
	SetPIIRules([]PIIRule{
		{Path: []string{"uid"}, Mode: PIIModeHash},
	})
	payload := map[string]any{"uid": "user-42"}
	result := SanitizePayload(payload, true, 32)
	hashed, ok := result["uid"].(string)
	if !ok {
		t.Fatalf("expected string hash value, got %T", result["uid"])
	}
	if len(hashed) != 12 {
		t.Errorf("expected 12 hex chars, got %q (len %d)", hashed, len(hashed))
	}
	// Verify it's hex.
	for _, c := range hashed {
		if !strings.ContainsRune("0123456789abcdef", c) {
			t.Errorf("non-hex char %q in hash %q", c, hashed)
		}
	}
}

// ── Test 6: custom rule — truncate mode ──────────────────────────────────────

func TestSanitizePayload_CustomRule_Truncate_String(t *testing.T) {
	resetPII(t)
	SetPIIRules([]PIIRule{
		{Path: []string{"note"}, Mode: PIIModeTruncate, TruncateTo: 5},
	})
	payload := map[string]any{"note": "hello world"}
	result := SanitizePayload(payload, true, 32)
	if result["note"] != "hello..." {
		t.Errorf("expected %q, got %v", "hello...", result["note"])
	}
}

func TestSanitizePayload_CustomRule_Truncate_ShortString(t *testing.T) {
	resetPII(t)
	SetPIIRules([]PIIRule{
		{Path: []string{"note"}, Mode: PIIModeTruncate, TruncateTo: 20},
	})
	payload := map[string]any{"note": "hi"}
	result := SanitizePayload(payload, true, 32)
	if result["note"] != "hi" {
		t.Errorf("expected %q unchanged, got %v", "hi", result["note"])
	}
}

func TestSanitizePayload_CustomRule_Truncate_NonString_Stringified(t *testing.T) {
	resetPII(t)
	SetPIIRules([]PIIRule{
		{Path: []string{"count"}, Mode: PIIModeTruncate, TruncateTo: 5},
	})
	payload := map[string]any{"count": 42}
	result := SanitizePayload(payload, true, 32)
	// fmt.Sprintf("%v", 42) = "42" (2 chars < 5) → no truncation
	if result["count"] != "42" {
		t.Errorf("expected %q for non-string truncate, got %v", "42", result["count"])
	}
}

// ── Test 7: wildcard * in path ────────────────────────────────────────────────

func TestSanitizePayload_WildcardPath(t *testing.T) {
	resetPII(t)
	SetPIIRules([]PIIRule{
		{Path: []string{"*", "secret_field"}, Mode: PIIModeDrop},
	})
	payload := map[string]any{
		"acct1": map[string]any{"secret_field": "s1", "name": _testAlice},
		"acct2": map[string]any{"secret_field": "s2", "name": "bob"},
	}
	result := SanitizePayload(payload, true, 32)

	acct1, _ := result["acct1"].(map[string]any)
	if acct1 == nil {
		t.Fatal("expected acct1 to be present")
	}
	if _, ok := acct1["secret_field"]; ok {
		t.Error("expected acct1.secret_field to be dropped")
	}
	if acct1["name"] != _testAlice {
		t.Errorf("expected acct1.name=%s, got %v", _testAlice, acct1["name"])
	}

	acct2, _ := result["acct2"].(map[string]any)
	if acct2 == nil {
		t.Fatal("expected acct2 to be present")
	}
	if _, ok := acct2["secret_field"]; ok {
		t.Error("expected acct2.secret_field to be dropped")
	}
}

// ── Test 8: nested map traversal ─────────────────────────────────────────────

func TestSanitizePayload_NestedMap_ThreeLevels(t *testing.T) {
	resetPII(t)
	payload := map[string]any{
		"level1": map[string]any{
			"level2": map[string]any{
				"password": "deep-secret", // pragma: allowlist secret
				"safe":     "ok",
			},
		},
	}
	result := SanitizePayload(payload, true, 32)

	l1, _ := result["level1"].(map[string]any)
	if l1 == nil {
		t.Fatal("expected level1")
	}
	l2, _ := l1["level2"].(map[string]any)
	if l2 == nil {
		t.Fatal("expected level2")
	}
	if l2["password"] != _piiRedacted { // pragma: allowlist secret
		t.Errorf("expected nested password redacted, got %v", l2["password"])
	}
	if l2["safe"] != "ok" {
		t.Errorf("expected safe unchanged, got %v", l2["safe"])
	}
}

// ── Test 9: slice traversal ───────────────────────────────────────────────────

func TestSanitizePayload_SliceOfMaps(t *testing.T) {
	resetPII(t)
	payload := map[string]any{
		"users": []any{
			map[string]any{"name": "alice", "token": "t1"},
			map[string]any{"name": "bob", "token": "t2"},
		},
	}
	result := SanitizePayload(payload, true, 32)

	users, _ := result["users"].([]any)
	if len(users) != 2 {
		t.Fatalf("expected 2 users, got %d", len(users))
	}
	for i, u := range users {
		m, _ := u.(map[string]any)
		if m == nil {
			t.Fatalf("user %d: expected map", i)
		}
		if m["token"] != _piiRedacted {
			t.Errorf("user %d: expected token redacted, got %v", i, m["token"])
		}
	}
}

// ── Test 10: depth limit enforced ────────────────────────────────────────────

func TestSanitizePayload_DepthLimit_StopsRecursion(t *testing.T) {
	resetPII(t)
	// With maxDepth=1, nested maps at depth 2 are not traversed.
	payload := map[string]any{
		"outer": map[string]any{
			"password": "should-not-be-redacted", // pragma: allowlist secret
		},
	}
	result := SanitizePayload(payload, true, 1)

	outer, _ := result["outer"].(map[string]any)
	if outer == nil {
		t.Fatal("expected outer map")
	}
	// At depth=1, recursion into outer does not happen, so password stays.
	if outer["password"] == _piiRedacted { // pragma: allowlist secret
		t.Error("expected password NOT redacted at depth limit of 1")
	}
}

// ── Test 11: maxDepth=0 uses default 8 ───────────────────────────────────────

func TestSanitizePayload_ZeroMaxDepth_UsesDefault(t *testing.T) {
	resetPII(t)
	// Build a 6-level deep structure with a password at the bottom.
	// Default depth is 8, so a 6-level nest must be fully sanitized.
	inner := map[string]any{"password": "deep"} // pragma: allowlist secret
	current := inner
	for i := 0; i < 5; i++ {
		current = map[string]any{"nested": current}
	}

	result := SanitizePayload(current, true, 0)
	// Traverse down to find the password.
	node := result
	for i := 0; i < 5; i++ {
		next, _ := node["nested"].(map[string]any)
		if next == nil {
			t.Fatalf("expected nested map at depth %d", i)
		}
		node = next
	}
	if node["password"] != _piiRedacted { // pragma: allowlist secret
		t.Errorf("expected deep password redacted, got %v", node["password"])
	}
}

// ── Test 12: SetPIIRules / GetPIIRules round-trip ────────────────────────────

func TestSetGetPIIRules_RoundTrip(t *testing.T) {
	resetPII(t)
	rules := []PIIRule{
		{Path: []string{"a", "b"}, Mode: PIIModeDrop},
		{Path: []string{"*", "x"}, Mode: PIIModeHash},
	}
	SetPIIRules(rules)
	got := GetPIIRules()

	if len(got) != len(rules) {
		t.Fatalf("expected %d rules, got %d", len(rules), len(got))
	}
	for i, r := range rules {
		if got[i].Mode != r.Mode {
			t.Errorf("rule %d: mode mismatch, want %q got %q", i, r.Mode, got[i].Mode)
		}
		if len(got[i].Path) != len(r.Path) {
			t.Errorf("rule %d: path length mismatch", i)
		}
	}
	// Verify isolation: mutating returned slice doesn't affect global.
	got[0].Mode = _testMutated
	got2 := GetPIIRules()
	if got2[0].Mode == _testMutated {
		t.Error("expected GetPIIRules to return independent copy")
	}
}

// ── Test 13: _resetPIIRules clears rules ─────────────────────────────────────

func TestResetPIIRules_ClearsRules(t *testing.T) {
	_resetPIIRules()
	t.Cleanup(_resetPIIRules)

	SetPIIRules([]PIIRule{
		{Path: []string{"x"}, Mode: PIIModeDrop},
	})
	_resetPIIRules()
	rules := GetPIIRules()
	if len(rules) != 0 {
		t.Errorf("expected empty rules after reset, got %d", len(rules))
	}
}

// ── Test 14: hash is deterministic ───────────────────────────────────────────

func TestSanitizePayload_Hash_Deterministic(t *testing.T) {
	resetPII(t)
	SetPIIRules([]PIIRule{
		{Path: []string{"uid"}, Mode: PIIModeHash},
	})
	payload := map[string]any{"uid": "user-99"}
	r1 := SanitizePayload(payload, true, 32)
	r2 := SanitizePayload(payload, true, 32)

	h1, _ := r1["uid"].(string)
	h2, _ := r2["uid"].(string)
	if h1 != h2 {
		t.Errorf("hash not deterministic: %q vs %q", h1, h2)
	}
}

// ── pii.go:142 depth-1 in map recursion ──────────────────────────────────────

func TestSanitizePayload_MapDepth_CorrectDecrement(t *testing.T) {
	resetPII(t)
	payload := map[string]any{
		"level1": map[string]any{
			"level2": map[string]any{
				"password": "deep-hidden", // pragma: allowlist secret
			},
		},
	}
	result := SanitizePayload(payload, true, 2)

	l1, _ := result["level1"].(map[string]any)
	if l1 == nil {
		t.Fatal("expected level1")
	}
	l2, _ := l1["level2"].(map[string]any)
	if l2 == nil {
		t.Fatal("expected level2")
	}
	if l2["password"] == _piiRedacted { // pragma: allowlist secret
		t.Error("expected password NOT redacted when map depth limit reached (depth=1 stops recursion)")
	}
}

// ── pii.go:144 depth-1 vs depth+1 in _sanitizeSlice ─────────────────────────

func TestSanitizePayload_SliceDepth_CorrectDecrement(t *testing.T) {
	resetPII(t)
	payload := map[string]any{
		"arr": []any{
			map[string]any{
				"nested": map[string]any{
					"password": "deep-secret", // pragma: allowlist secret
				},
			},
		},
	}
	// maxDepth=3: arr(depth=3) → slice(depth=2) → map(depth=2) → nested(depth=2>1 recurse)
	// → map(depth=1) → password → sensitive → redacted
	result := SanitizePayload(payload, true, 3)

	arr, _ := result["arr"].([]any)
	if arr == nil {
		t.Fatal("expected arr in result")
	}
	item, _ := arr[0].(map[string]any)
	if item == nil {
		t.Fatal("expected map item in arr")
	}
	nested, _ := item["nested"].(map[string]any)
	if nested == nil {
		t.Fatal("expected nested map")
	}
	if nested["password"] != _piiRedacted { // pragma: allowlist secret
		t.Errorf("expected deep password redacted at depth 3, got %v", nested["password"])
	}

	// Now verify depth boundary: maxDepth=2 stops before reaching the password.
	result2 := SanitizePayload(payload, true, 2)
	arr2, _ := result2["arr"].([]any)
	item2, _ := arr2[0].(map[string]any)
	// At depth=2: _sanitizeSlice called with depth=2-1=1.
	// Inner map processed at depth=1. depth<=1 → no recursion into nested map.
	// nested map is returned unchanged.
	nested2, _ := item2["nested"].(map[string]any)
	if nested2 == nil {
		t.Fatal("expected nested map at depth 2")
	}
	if nested2["password"] == _piiRedacted { // pragma: allowlist secret
		t.Error("expected password NOT redacted when depth stops before reaching it")
	}
}
